From 0b1c34eef47116580fb389dcf2013cdaa28c346b Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jdapena@igalia.com>
Date: Tue, 17 Dec 2019 16:18:23 +0100
Subject: libweston: release touch grab on touch up

Chromium relies on dnd_finished event to free the data source. But weston
will only emit the signal and process the wl_data_offer finish() if the
drag operation has been complete. This happens as expected in mouse drop, but
equivalent code is missing in touch_up handler.
---
 libweston/data-device.c | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/libweston/data-device.c b/libweston/data-device.c
index e3dbee3e..751d0c67 100644
--- a/libweston/data-device.c
+++ b/libweston/data-device.c
@@ -743,14 +743,32 @@ drag_grab_touch_up(struct weston_touch_grab *grab,
 	struct weston_touch_drag *touch_drag =
 		container_of(grab, struct weston_touch_drag, grab);
 	struct weston_touch *touch = grab->touch;
+	struct weston_data_source *data_source = touch_drag->base.data_source;
 
 	if (touch_id != touch->grab_touch_id)
 		return;
 
-	if (touch_drag->base.focus_resource)
-		wl_data_device_send_drop(touch_drag->base.focus_resource);
-	if (touch_drag->base.data_source)
+	if (data_source) {
+		if (touch_drag->base.focus_resource &&
+		    data_source->accepted &&
+		    data_source->current_dnd_action) {
+			wl_data_device_send_drop(touch_drag->base.focus_resource);
+
+			if (wl_resource_get_version(data_source->resource) >=
+			    WL_DATA_SOURCE_DND_DROP_PERFORMED_SINCE_VERSION)
+				wl_data_source_send_dnd_drop_performed(data_source->resource);
+
+			data_source->offer->in_ask =
+				data_source->current_dnd_action ==
+				WL_DATA_DEVICE_MANAGER_DND_ACTION_ASK;
+
+			data_source->seat = NULL;
+		} else if (wl_resource_get_version(data_source->resource) >=
+			   WL_DATA_SOURCE_DND_FINISHED_SINCE_VERSION) {
+			wl_data_source_send_cancelled(data_source->resource);
+		}
 		wl_list_remove(&touch_drag->base.data_source_listener.link);
+	}
 	data_device_end_touch_drag_grab(touch_drag);
 }
 
-- 
2.20.1

